src/utils/mutate.js

```diff
@@ -0,0 +1,249 @@
+import { chunk, cloneDeep, flatten, mapValues } from 'lodash';
+import { firestoreRef } from './query';
+import mark from './profiling';
+
+/**
+ * @param {object} firestore
+ * @param {string} collection
+ * @param {string} doc
+ * @returns Boolean
+ */
+const docRef = (firestore, collection, doc) =>
+  firestore.doc(`${collection}/${doc}`);
+
+/**
+ * @param object
+ * @returns Promise
+ */
+async function promiseAllObject(object) {
+  return Object.fromEntries(
+    await Promise.all(
+      Object.entries(object).map(([key, promise]) =>
+        promise.then((value) => [key, value]),
+      ),
+    ),
+  );
+}
+
+/**
+ * @param {object} operations
+ * @returns Boolean
+ */
+function isBatchedWrite(operations) {
+  return Array.isArray(operations);
+}
+
+/**
+ * @param {Mutation_v1 | Mutation_v2} read
+ * @returns Boolean
+ */
+function isDocRead(read) {
+  return read && typeof read.doc === 'string';
+}
+
+/**
+ * @param {Mutation_v1 | Mutation_v2} operations
+ * @returns Boolean
+ */
+function isSingleWrite(operations) {
+  return operations && typeof operations.collection === 'string';
+}
+
+// ----- FieldValue support -----
+
+const primaryValue = (arr) =>
+  Array.isArray(arr) && typeof arr[0] === 'string' && arr[0].indexOf('::') === 0
+    ? null
+    : arr;
+
+const arrayUnion = (firebase, key, val) => {
+  if (key !== '::arrayUnion') return null;
+  return firebase.firestore.FieldValue.arrayUnion(val);
```

Comment:
This may be the answer to my question as to why we need the string keys - I'm wondering though.... could those maybe just be passed along like so:

```js
// changes requested to your data
const firestoreDocument = {
  otherNumber: firebase.firestore.FieldValue.increment(5)
}
```

Not sure it is even necessary, but if they needed to be dynamic to the val, we could maybe use a function like:

```js
// changes requested to your data
const firestoreDocument = {
  otherNumber: (val) => firebase.firestore.FieldValue.increment(val),
}
```

Reply:
Ahhh.... I think I finally see... Is it because these are going into the store and we need those to remain as plain objects?

Reply:
Right. The value is needed to have the optimistic value in the store. But inspecting the inside of a FieldValue isn’t good. If/when Firestore changes the internals of their code the hacky way to get the value would be different. 

I’ll clean up the docs so this is clear to consumers of the library.